Use is_downloaded for load and download

edwardleardi opened the pull request:
Add a force_check param to Dataset load and download methods
Fix a bug with _file_list_file

---

## Checklist:

<details open>

<summary></summary>

- Does this pull request close an issue? We encourage you to open an issue first if this pull request (PR) is not a
  minor change.
  - [ ] No
    - [ ] The change in this pull request is minor.
  - [x] Yes
    - Close #76

For the following questions, only check the boxes that are applicable.

- [x] Ensure that the pull request text above the horizontal line is descriptive.
- [x] Add/update relevant tests.
- [x] Add/update relevant documents.

- Do you have triage permission of this repository?
  - [ ] No
    - You are good.
  - [x] Yes
    - Does this pull request close an issue?
      - [ ] No, this pull request also acts as an issue by itself.
        - [ ] Assign yourself.
        - [ ] Label this pull request.
        - [ ] Put a milestone.
        - [ ] Put this pull request under "In Progress" or "Under Review" pipeline on ZenHub.
        - [ ] Put this pull request under the appropriate epic on ZenHub.
        - [ ] Put an estimate on ZenHub.
        - [ ] Add dependency relationship with other issues/pull requests on ZenHub.
      - [x] Yes, this pull request addresses an issue and does not act as an issue.
        - [x] Connect this pull request to the underlying issue on ZenHub.
        - [x] DON'T:
          - DON'T assign yourself or anyone else.
          - DON'T label.
          - DON'T put a milestone.
          - DON'T put this pull request under any epic on ZenHub.
          - DON'T put an estimate on ZenHub.
    - [x] When the pull request is ready, request review.
</details>

edwardleardi committed:
Use is_downloaded for load and download

Add a force_check param to Dataset load and download methods
Fix a bug with _file_list_file

edwardleardi commented on pydax/_dataset.py:
```diff
@@ -119,17 +119,32 @@ def _pydax_dir(self) -> pathlib.Path:
-        return self._pydax_dir / 'files.list'
+        # We use self._pydax_dir_ instead of self._pydax_dir because we don't want to have the directory created if it
+        # hasn't been yet when is_downloaded is called
+        return self._pydax_dir_ / 'files.list'
```
Was this a bug?

xuhdev replied:
I don't think this is a bug by itself. Seems to me the best way to fix the issue is to create another property that ends with `_` that doesn't create the directory structure.

edwardleardi replied:
Is there a situation though where `_file_list_file` needs to call `_pydax_dir` over `_pydax_dir_`? 

In `Dataset.download`, by the time `_file_list_file` is called, `_pydax_dir` was already called earlier. In `Dataset.is_downloaded` we just need to check if `_file_list_file` exists so we don't need `_pydax_dir` to have been called.

xuhdev replied:
`is_downloaded` is where I think it should be called

edwardleardi replied:
can you elaborate a little bit on why though?

xuhdev replied:
It's just that the directory shouldn't be created in `is_downloaded`. The directory shouldn't be created merely because `is_downloaded` returns False.

edwardleardi replied:
> It's just that the directory shouldn't be created in is_downloaded. The directory shouldn't be created merely because is_downloaded returns False.

I think one of us is confused then, because I agree and that's the change I am proposing

edwardleardi committed:
Fix bug in tests due to _file_list_file

Complement _file_list_file with _file_list_file_
Fix ensuing bug with TemporaryDirectory in tests that use Dataset.delete

xuhdev commented on pydax/_dataset.py:
```diff
@@ -95,7 +95,10 @@ def __init__(self, schema: SchemaDict,
+        "Same as :attr:`_data_dir_`, but create it if it does not exist."
+        # We use a convention for attribute pairs that return the same pathlib.Path where a trailing "_" indicates the
+        # path should be used for reading data (e.g. checking the path exists) while no trailing "_" indicates the path
+        # is for writing data (e.g. creating the file)
```
How about moving these comments below the class docstring?

xuhdev commented on pydax/_dataset.py:
```diff
@@ -109,27 +112,45 @@ def _pydax_dir_(self) -> pathlib.Path:
 
-        :raises NotADirectoryError: :attr:`Dataset._data_dir` (passed in via ``data_dir`` in the constructor
-                                    :class:`Dataset`) points to an existing file that is not a directory.
+        :param force_check: Force a check that the data files were not previously downloaded to
```
How about renaming `force_check` to `check`? You default it to `True`, which doesn't sound like forcing.

edwardleardi replied:
sure thing

xuhdev commented on pydax/_dataset.py:
```diff
@@ -166,19 +187,31 @@ def download(self) -> None:
         :param format_loader_map: The :class:`.loaders.FormatLoaderMap` object that determines which loader to use.
-        :raises FileNotFoundError: The dataset files are not found on the disk. Usually this is because
-            :func:`~Dataset.download` has never been called.
+        :param force_check: Force a check that the data files were previously downloaded to :attr:`Dataset._data_dir`
```
Same rename suggestion as above.

xuhdev commented on pydax/_dataset.py:
```diff
@@ -223,24 +256,30 @@ def data(self) -> Dict[str, Any]:
+        .. warning::
+
+            :meth:`.is_downloaded` will search for the dataset files in :attr:`Dataset._data_dir` (passed in via
+            ``data_dir`` in the constructor :class:`Dataset`). If after downloading, you move the data files to a
```
```suggestion
            ``data_dir`` in the constructor :class:`Dataset`). If after downloading, you manipulate the data files outside the control of this library
```

xuhdev commented on pydax/_dataset.py:
```diff
@@ -166,19 +188,31 @@ def download(self) -> None:
+        if force_check and not self.is_downloaded():
+            raise RuntimeError(f'Failed to locate downloaded data files. Either {self.__class__.__name__}.download() '
+                               f'was not run or the data files were moved from {self.__class__.__name__}._data_dir '
+                               f'(passed in via data_dir in the constructor {self.__class__.__name__}).')
```
How about we make this more concise:

> Downloaded data files are not present in {self.data_dir} or corrupted.

edwardleardi replied:
good idea

edwardleardi committed:
Apply suggestions from code review

Co-authored-by: Hong Xu <hongx@ibm.com>

xuhdev commented on pydax/_dataset.py:
```diff
@@ -109,27 +113,43 @@ def _pydax_dir_(self) -> pathlib.Path:
-        :raises NotADirectoryError: :attr:`Dataset._data_dir` (passed in via ``data_dir`` in the constructor
-                                    :class:`Dataset`) points to an existing file that is not a directory.
+        :param check: Check that the data files were not previously downloaded to :attr:`Dataset._data_dir` (passed in
+            via ``data_dir`` in the constructor :class:`Dataset`) by running :meth:`.is_downloaded`.
```
There's a slight ambiguity here. What's the effect of the check being positive/negative?

xuhdev commented on pydax/_dataset.py:
```diff
@@ -109,27 +113,43 @@ def _pydax_dir_(self) -> pathlib.Path:
+            via ``data_dir`` in the constructor :class:`Dataset`) by running :meth:`.is_downloaded`.
+        :raises RuntimeError: The dataset was previously downloaded as indicated by :meth:`.is_downloaded`
+            returning ``True``.
+        :raises NotADirectory: :attr:`Dataset._data_dir` (passed in via ``data_dir`` in the constructor
```
```suggestion
        :raises NotADirectoryError: :attr:`Dataset._data_dir` (passed in via ``data_dir`` in the constructor
```

xuhdev commented on pydax/_dataset.py:
```diff
@@ -166,19 +186,28 @@ def download(self) -> None:
         :param format_loader_map: The :class:`.loaders.FormatLoaderMap` object that determines which loader to use.
-        :raises FileNotFoundError: The dataset files are not found on the disk. Usually this is because
-            :func:`~Dataset.download` has never been called.
+        :param check: Check that the data files were previously downloaded to :attr:`Dataset._data_dir` (passed in via
```
Same as above

xuhdev commented on pydax/_dataset.py:
```diff
@@ -109,27 +113,45 @@ def _pydax_dir_(self) -> pathlib.Path:
+        :param check: Check to make sure the data files are not already present in :attr:`._data_dir` (passed in via
+            ``data_dir`` in the constructor :class:`Dataset`) by running :meth:`.is_downloaded`. Raise an error if
+            they are and prevent a subsequent download. Set to ``False`` to avoid this safeguard. Subsequent calls to
+            :meth:`.download` will then overwrite data files if they were previously downloaded to :attr:`._data_dir`.
```
```suggestion
            ``data_dir`` in the constructor :class:`Dataset`) by running :meth:`.is_downloaded`. If set to True, raise an error if
            they are present and prevent a subsequent download. Set to ``False`` to remove this safeguard, and subsequent calls to
            :meth:`.download` will then overwrite data files if they were previously downloaded to :attr:`._data_dir`.
```
